### ISSUE TITLE
Extend survival analysis

### ISSUE THREAD

**Issue by PhilippPro** [mlr-org/mlr#1353]
I am just starting to get to know survival analysis and I see some things we could possibly extend. 

For example it is usual to look at (e.g. cross-validated) prediction error curves, like in the pec package. 
Moreover we could extend the measure section. At the moment there is only the concordance index, but a typical measure is also the integrated brier score (which is based on the prediction error curve). This is also provided by the pec package (for some learners).

I think at the moment I would rather use the pec package than the mlr package for analysing survival predictions. 

This paper is also quite interesting and extends the pec package to other models like random forest: 
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4194196/

This issue is of course related to #1321.

**Comment by berndbischl** [comment]
@PhilippPro @mllg 
can you guys talk about this to each other?

**Comment by PhilippPro** [comment]
Ok, so for estimating the pec (prediction error curvee) it is necessary to obtain a matrix with probabilities at different points in time for each observation and a vector that specifies the points in time. With this information the pec can be calculated and hence the IBS. 

This matrix is provided e.g. by the randomForestSRC package (from their help):

```
# primary call
set.seed(542899)
data(pbc, package = "randomForestSRC")
train <- sample(1:nrow(pbc), round(nrow(pbc) * 0.50))
pbc.out <- rfsrc(Surv(days, status) ~ .,  data=pbc[train, ],
                 nsplit = 10)

# standard predict call
pbc.train <- predict(pbc.out, pbc[-train, ], outcome = "train")
Time <- pbc.out$time.interest
matplot(Time, t(exp(-pbc.train$chf)[1:3,]), ylab = "Survival", col = 1, type = "l")
```
So this matrix has to be stored in mlr. The aggregation e.g. in CV has to aggregate the "probability matrix" results of all the folds.

**Comment by mllg** [comment]
Correct me if I am wrong, but `pec` seems to need a matrix of probabilities for all time points in the test set. This is rather ugly:

* We need to access the test set during training
* We cannot aggregate matrices because they might have different dimensions
* We cannot predict on "new" data

We could use the `predictSurvProb` function to dynamically create survival predictions during `predict.[learner]`. Unfortunately, calculating the PECs is rather ugly:
```r
library(mlr)
library(survival)
library(pec)
data = getTaskData(wpbc.task)
train = sample(nrow(data), 2/3 * nrow(data))
test = setdiff(seq_len(nrow(data)), train)

# train.coxph:
mod = coxph(Surv(time, status) ~ ., data = data[train, ])

# predict.coxph:
probs = predictSurvProb(mod, newdata = data[test, ], times = data[test, "time"])

# nothing here works for me:
pec(probs)
pec(list(coxph = probs))
pec(list(coxph = probs), formula = Surv(time, status) ~ ., data = data[test, ])
pec(list(coxph = mod), formula = Surv(time, status) ~ ., data = data[test, ])

# this kind of works :/
f = as.formula(sprintf("Surv(time, status) ~ %s", paste(setdiff(names(data), c("time", "status")), collapse = "+")))
pec(list(coxph = mod), formula = f, data = data[test, ])
```

Any thoughts?

**Comment by PhilippPro** [comment]
> * We need to access the test set during training

I do not understand this. Why do you have to access it? Every necessary information (e.g. in randomForestSRC) is given in the prediction object. 
> * We cannot aggregate matrices because they might have different dimensions

They can be aggregated by building a new matrix with inserting all possible time points in all folds as row dimension. A bit complicated.
> * We cannot predict on "new" data

Why? As you can see in my example we get probabilities for new "test" data. 

My idea would be rather write our own "pec" function, because you can apply the pec function only on the trained model and not on the predicted model or predicted data. And internally pec does/can do some crossvalidation/bootstrap, which we also do. But I am also not 100 % clear about this, your last function seems also to be a possible approach.

**Comment by mllg** [comment]
> I do not understand this. Why do you have to access it? Every necessary information (e.g. in randomForestSRC) is given in the prediction object.

Whops, you might be right. I've mixed up the variables you used. However, your approach seems to only work for `rfsrc` and is no generic solution. Getting the survival probabilities for a larger number of models is maybe the most difficult part.

> They can be aggregated by building a new matrix with inserting all possible time points in all folds as row dimension. A bit complicated.

Yep, possible but complicated. We could also just aggregate the IBS of multiple train/test splits?

> My idea would be rather write our own "pec" function, because you can apply the pec function only on the trained model and not on the predicted model or predicted data. And internally pec does/can do some crossvalidation/bootstrap, which we also do. But I am also not 100 % clear about this, your last function seems also to be a possible approach.

This is the cleanest but also most laborious solution ;)

**Comment by PhilippPro** [comment]
> Yep, possible but complicated. We could also just aggregate the IBS of multiple train/test splits?

Yes, I also thought about this. I am not 100 % sure, but I think it yields the same results, at least in the non-censored case.

**Comment by studerus** [comment]
In the follwing example, it is not necessary to provide the originally fitted model to the `pec` function, only the predicted probabilities

```
timepoints = seq(0, max(data$time), length.out = 100L)
probs = predictSurvProb(mod, newdata = data[test, ], times = timepoints)

pec(probs, Surv(time, status) ~ 1, data = data[test,], exact = F,
    exactness = 99L, maxtime = max(timepoints))
```

We could use a grid of equidistant time points for prediction that stays the same over all cross validation folds. For simplicity, I used `Surv(time, status) ~ 1` which has the effect, that the Kaplan-Meier estimator is used for estimating the censoring weights. One potential problem here is that the conditional censoring model has to be estimated repeatedly from the testing data at each fold. However, if cross validation using the pec function is performed, it seems to be estimated from the whole data set. See description of the argument `ipcw.refit` (If TRUE the inverse probability of censoring weigths are estimated separately in each training set during cross-validation). This is true by default, which makes sense to me because test data sets during cross validation might be too small to reliably estimate this model. Besides, it would violate the principle of strictly separating testing from learning.

**Comment by studerus** [comment]
Sorry, `ipcw.refit` is actually `FALSE` by default.

**Comment by studerus** [comment]
I think, we should also add time dependent AUC measures. This paper just came out:
[A simple method to estimate the time-dependent receiver operating characteristic curve and the area under the curve with right censored data](http://smm.sagepub.com/content/early/2016/11/25/0962280216680239.abstract?rss=1)
Their proposed method is implemented in the `tdROC` package.

We could create a time dependent AUC measure like this (here the time point is arbitrarily fixed to 2):
```
measureTDROC = function(task, model, pred, feats, extra.args) {
   s = getPredictionTruth(pred)
   X = getPredictionResponse(pred)
   tdROC::tdROC(X = X, Y = s[,1L], delta = s[,2L], tau = 2)$AUC$value
}

tdroc = makeMeasure(
  id = "tdROC",
  name = "Time-dependent AUC estimated using kernel weighting method",
  properties = c("surv", "req.pred", "req.truth"),
  minimize = FALSE, best = 1, worst = 0,
  fun = measureTDROC
)
```

The measure has several parameters, most importantly, the time at which the AUC should be estimated. How can we pass such parameters to a measure in `mlr`?

**Comment by mllg** [comment]
> How can we pass such parameters to a measure in mlr?

@berndbischl correct me if I'm wrong here.

This is currently not supported without creating your own measure using `makeMeasure`. For some (historic?) reasons, measures are objects and not functions. However, there is an easy undocumented way:

1) Implement a measure where the function signature contains the argument `extra.args`.
2) Copy the measure and set the parameter: `tdroc.10 = tdroc; tdroc.10$extra.args = list(t = 10)`.
3) Use `extra.args$t` in your measure function.

**Comment by studerus** [comment]
Thanks. I think we should have `setMeasurePars` and `getMeasurePars` functions to make this more user friendly.

**Comment by mllg** [comment]
We should maybe schedule a short call to agree on:

1) What measures do we want to have implemented. PEC/Brier, C-indices, AUC/ROC, ...

2) What information needs to be attached to the learner models to calculate the measures. All seem to need some extra information from training, but our aim should be to reduce this to a minimum.

**Comment by PhilippPro** [comment]
Would be good, maybe also clear who wants to do what. When do you (mllg, studerus) have time? I am quite flexible, no restrictions this week. (except of wednesday 16-17 o'clock)

**Comment by studerus** [comment]
Friday during the day or evening would be best for me.

**Comment by mllg** [comment]
How about Friday, 15:00?

**Comment by studerus** [comment]
ok

**Comment by PhilippPro** [comment]
For me also. See you on hangout.

**Comment by csetraynor** [comment]
Hi, 
Thank you for the tips presented here, they are extremely helpful. I wonder if all these functions have been already implemented or you need more workforce? I volunteer to help in anything,

**Comment by PhilippPro** [comment]
As you can see in the tutorial (https://mlr-org.github.io/mlr-tutorial/devel/html/measures/index.html), we have implemented in total 4 measures. They possibly should work correctly. ;)

**Comment by IyarLin** [comment]
Hi, first I'd like to thank you for the great work done on this. I've transitioned to using mlr instead of caret mainly due to it's great support for survival analysis, but also because it lookks better structured and maintained.

I do have one request: Can we add an option to predict survival curves? I imagine they are being calculated for the ibrier score anyway, so just make it possible to access them in the prediction object. I could use predictSurvProb on the model object nested within the trained model, but I don't see how I can apply wrappings to the target dataset as well.

Thanks!

Iyar

**Comment by PhilippPro** [comment]
You mainly have to thank @mllg. The visualization of survival tasks/predictions is not done at all, at the moment. Would be a very useful thing, I think. 

What you currently can do is to take the code in measures.R and adapt it to your needs:

```
targets = getTaskTargets(task)
tn = getTaskTargetNames(task)
f = as.formula(sprintf("Surv(%s, %s) ~ 1", tn[1L], tn[2L]))
newdata = getTaskData(task)[model$subset, ]
max.time = extra.args$max.time %??% max(newdata[[tn[1L]]])
grid = seq(0, max.time, length.out = extra.args$resolution)
probs = predictSurvProb(model$learner.model, newdata = newdata, times = grid)
```

**Comment by IyarLin** [comment]
If the learner is wrapped with preprocessing steps than the above model$learner.model object would be of class "WrappedModel" which is not accetable by predictSurvProb function. You can go further down the rabbit hole and fetch the actual model: model$learner.model$next.model$learner.model but than I don't think it applies the pre-processing steps

**Comment by PhilippPro** [comment]
Yes, I see your point. The problem is, that predictSurvProb is from the pec package and not from mlr, so it is not applicable on a wrapped model. We would need to write our own function for that.
